### PR TITLE
DAOS-11043 common: Blob backed Memory(BMEM) interface for umem

### PR DESCRIPTION
**Pull request by sherintg**
This commit contains a new allocator which is a modified version of pmemobj. The new allocator operates on volatile memory and provides transaction functionality similar to that of pmemobj.

The BMEM interfaces internally use the new allocator functionalties to implement the required backend for hosting metadata on volatile memory backed by blob.

Note: The current commit default to BMEM instead of PMDK. One can override this by setting DAOS_USE_PMDK=1 in the server config file.

Allow-unstable-test: true

Signed-off-by: sherintg <sherin-t.george@hpe.com>

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it
      should be force landed and forced-landing tag is set.
* [ ] Commit messages follows the guidelines outlined
      [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code.
* [ ] Any appropriate watchers have had a chance to review the PR.
* [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested
      changes.
* [ ] Githooks were used. If not, there is a sufficient reason to move forward.  Check copyrights
      if githooks were not used.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future
      PRs.
* [ ] All builds have passed.  Check non-required builds to ensure they don't indicate any problem
      such as a compiler warning on extraneous platforms.
* [ ] No new NLT or valgrind warnings.  Check the classic view. This step should only matter if
      the build requires force landing.
* [ ] Ensure sufficent testing is done.   Check feature pragmas and test tags. Check that tests
      skipped for the ticket are run and now pass with the changes.
* [ ] Quick-build or Quick-functional is not used.
* [ ] Pay attention to PRs that may affect compatibility between versions and ensure it has been
      addressed.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If
      it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Check the commit message when landing. Check the standard
      [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit
      it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

**Comment by jolivier23** [comment]
> Bug-tracker data: Ticket title is 'Implement new alternative umem backend' Status is 'In Progress' Labels: 'md_on_ssd' Job should run at elevated priority (3) Errors are Unknown component https://daosio.atlassian.net/browse/DAOS-11043

It doesn't like umem. You can either change it to another directory name under src/ or src/client or add umem as a component in VALID_COMPONENTS in ci/jira_query.py

This is not required but if you push again, this will eliminate the noise.

File: src/common/dav/alloc.h [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+static inline void *
+Realloc(void *ptr, size_t size)
+{
+	void *new_ptr;
+
+	D_REALLOC_NZ(new_ptr, ptr, size);
+	return new_ptr;
+}
+
+#define	Free(ptr)	D_FREE(ptr)
```
Comment (jolivier23) [inline src/common/dav/alloc.h:36]: One side effect and you can test with D_LOG_MASK=debug DD_MASK=mem is you may get logging for a variable named "ptr" in Malloc but the name will be whatever is passed to the macro here.  That shouldn't cause any issue but may be a little confusing to the log reader.   In such case, it may be better to wrap in an inline function.   I presume the compiler will make it efficient.

What I had in mind with my comment on the prior patch was to wrap the routines, rather than the underlying malloc/free/realloc simiilar to what we do with D_STRNDUP or D_ASPRINTF...the idea being that we log these routines differently.   I'm honestly fine either way though, just clarifying.

**Review by jolivier23 — COMMENTED** [review]
You might also check if NLT is doing any fault injection.  That could cause "new" failures in some places.

File: src/common/dav/alloc.h (hunk 1 above)
Comment (sherintg) [reply src/common/dav/alloc.h]: Even the NLT failures were a result of inconsistent use of D_ALLOC/* macros. There were no real double free getting called. It was just that D_FREE was called on allocations done using posix_mem_align() and strdup(). I have made a new commit where all of the dav routines will now use the DAOS D_ALLOC/FREE/* macros for internal allocation.

File: src/vos/tests/vts_ts.c [hunk 2]
```diff
@@ -474,6 +474,7 @@ lru_array_stress_test(void **state)
 		assert_int_equal(evicted, LRU_ARRAY_SIZE);
 	}
 
+	srand(1); /* The below test suite may fail for random seed value*/
```
Comment (jolivier23) [inline src/vos/tests/vts_ts.c:477]: why is this necessary here?   Do we use random elsewhere in the code differently than before?

**Comment by jolivier23** [comment]
I merged latest base branch and tested the failing core tests.  I wasn't able to reproduce the rebuild failure

File: src/vos/tests/vts_ts.c (hunk 2 above)
Comment (sherintg) [reply src/vos/tests/vts_ts.c:477]: This particular test  sometimes fails if the random number seed is different. During vos create the new code is calling uuid_create() which inturn is using rand. This inturn is causing this test suite to fail sometimes as the below for loop expects 70:30 proportion of lru_alloc vs lru_evict calls to be made and sometimes it is not exactly met.

Comment (jolivier23) [reply src/vos/tests/vts_ts.c:477]: Ok, It's probably fine but I tried this several times with and without the envirable and it doesn't crash.

**Review by jolivier23 — APPROVED** [review]
